### DISCUSSION TITLE
How to rotate Root CA for running self-hosted step-ca server ?

### DISCUSSION THREAD

**Discussion by chris-lee-lb** [smallstep/certificates#1154] [Q&A] [answered]
Hello,

As title, I want to figure out the correct steps for it. After some googling, I still can not found any information about it.

I do found this official blog post - https://smallstep.com/blog/step-v0.8.3-federation-root-rotation/

But as my understanding, It says I can use federation (multiple Root CAs) to do seamlessly rotate root certificates, but do not have any informations that covers root rotations.

Second thoughts is that maybe I can use step-ca JSON API, but I also can not find any docs for json api endpoint usage. 

Thanks in advance.

**Comment by chris-lee-lb** [comment, marked as answer]
Currently I found this issue and related PR for this topic.

https://github.com/smallstep/certificates/issues/19
https://github.com/smallstep/certificates/pull/22

Looks like step ca do have related mechanism with following steps.

1. CA is configured with rootA.
2. BootstrapServer/BootstrapClient are created, accepting only rootA.
3. CA is updated and adds rootB, rootA and rootB are accepted.
4. BootstrapServer/BootstrapClient do a renew and include rootB into the trusted pools
5. CA can now create new intermediates and remove rootA.
6. BootstrapServer/BootstrapClient will able to renew with rootB.

But I still can not figure out how to do these steps actually.

Thanks in advance.

**Reply by maraino** [reply]
In the `ca.json` you can configure multiple roots like:

```json
{
  "root": ["/path/to/rootA.crt", "/path/to/rootB.crt"],
  "...": "..."
}
```

This configuration allows the CA to trust all certificates derived from rootA and rootB. The command `step ca roots` will download both. 

If we want a client using the old root, let's say rootB.crt to renew a certificate, we will need to:

1. Replace the clients $(step path)/certs/root_ca.crt with either the new root or the bundle created by `step ca roots`. 
2. Renew as usual. 

The renewed cert will be signed by the intermediate used in the rootA PKI.

Once all new certificates are issued by rootA, you can get rid of rootB.

**Reply by chris-lee-lb** [reply]
@maraino Thanks for your reply. I'm glad that I have same steps after reading your PR and some googling.

But because I use step ca with autocert via k8s helm charts. Currently have two critical problems to do.

(1) step-ca helm chart do not support that have multiple root CAs.
(2) autocert do not keep sync root CAs. (also not support have multiple root CAs)

But still thanks your advice. 🎉

**Reply by maraino** [reply]
We should be able to configure step-ca helm chart with two roots, but I will have to look into that it might not be straightforward.

**Reply by maraino** [reply]
Right now you cannot manually change the output of `step ca init --helm` to configure two roots, but you can manually edit the certs and config ConfigMaps to add the second root.

I think I will add a new change, so the second certificate is not required and you can just bundle the old root with the new one. This way you will need is to add it to the output of `step ca init --helm`

**Reply by maraino** [reply]
@chris-lee-lb I've added a PR to support root bundles https://github.com/smallstep/certificates/pull/1169, so in the output of `step ca init --helm` you can append multiple roots :)

**Reply by maraino** [reply]
Autocert pods will either require a rollout restart. The other option would be to configure custom bootstrappers and/or renewer containers and add an extra command to get all the roots.

**Reply by chris-lee-lb** [reply]
@maraino Thanks for your nice works. 🎉  

And for Autocert, I also create the related issue for this case. (https://github.com/smallstep/autocert/issues/67), and I also have some idea in the comments. Hope it be helpful.

**Comment by ken-master** [comment]
hi @chris-lee-lb were you able to implement and configure these stuff.. I am in a same situation. Thanks.
